dropshot/src/server.rs

```diff
@@ -257,17 +252,41 @@ impl<C: ServerContext> InnerHttpServerStarter<C> {
     /// Begins execution of the underlying Http server.
     fn start(
         self,
-        close_signal: tokio::sync::oneshot::Receiver<()>,
+        mut close_signal: tokio::sync::oneshot::Receiver<()>,
         log_close: Logger,
-    ) -> tokio::task::JoinHandle<Result<(), hyper::Error>> {
-        let graceful = self.0.with_graceful_shutdown(async move {
-            close_signal.await.expect(
-                "dropshot server shutting down without invoking close()",
-            );
-            info!(log_close, "received request to begin graceful shutdown");
-        });
+    ) -> tokio::task::JoinHandle<()> {
+        use hyper_util::rt::{TokioExecutor, TokioIo, TokioTimer};
+        use hyper_util::server::conn::auto;
+
+        tokio::spawn(async move {
+            let mut builder = auto::Builder::new(TokioExecutor::new());
+            // http/1 settings
+            builder.http1().timer(TokioTimer::new());
+
+            let graceful =
+                hyper_util::server::graceful::GracefulShutdown::new();
 
-        tokio::spawn(graceful)
+            loop {
+                tokio::select! {
+                    Ok((sock, remote_addr)) = self.0.accept() => {
+                        let fut = builder.serve_connection_with_upgrades(
+                            TokioIo::new(sock),
+                            self.1.make_svc(remote_addr),
+                        );
+                        let fut = graceful.watch(fut.into_owned());
```

Comment:
I'm having a hard time understanding the various futures involved here and how the new `Graceful` mechanism works.  Would it be reasonable to add a comment describing this in more detail?  And/or using more descriptive names?

Reply:
I added some comments.